### DISCUSSION TITLE
Custom Column Collapsible... refresh only if visible.

### DISCUSSION THREAD

**Discussion by epertinez** [filamentphp/filament#8878] [General]
Hi, I don't know if this is a new feature or there is a clever way to do it, but I feel it may be a common problem.

I have this collapsible and collapsed CustomColumn that expands Column.

This column is very expensive, in time and resources... SO I DON'T WANT THE COLUMN TO BE DOWNLOADED EXCEPT IF IT IS VISIBLE. And... i don't mind the user to wait while the column is loaded.

Now... I found a way to implement defered/lazy loading of the column through saving in a private attribute in the page/widget the column and then creating a method in the widget/page that calls the column method.

Yet... even in deffered/lazy loading... all column appearences get loaded at the same time, being hidden or not.

That takes too much.

Did not find a way to make the column aware of its visibility, so it only gets the expensive data if it is being shown. Livewire's php part does not seem to know about the DOM state at the time  it is being called.

I did nor find a way to make the column ONLY refresh a cell (that which button to show has been pushed) and not all other cells that remain invisible.

I think it is an important thing. Please, indicate me what am I doing wrong,  because it is hard to believe I am the only one with this problem.

**Comment by VincvanAs** [comment]
Could you share some code?

About your way of defer loading, have you seen/tried ->deferLoading(); ?

**Comment by epertinez** [comment]
Here you have a version of my code.
My Page has two Widgets. One of them is a table:
The Widget:

```
use App\Tables\Columns\CompetidorsTaulaDetall;

class CompetidorsTaula extends BaseWidget
{
    protected static ?string $pollingInterval = null;
    
    private $competidorsTaulaDetall;

    public function loadDetails() {
        $this->competidorsTaulaDetall->loadDetails();
    }
    
    public function table(Table $table): Table
    {  $this->competidorsTaulaDetall=CompetidorsTaulaDetall::make('ampliacio');
       return $table
            ->query(
                // ...
            )
            ->columns([
                // ...
                Grid::make(8)->schema(/* a bunch of columns */) ,
                Panel::make([
                    $this->competidorsTaulaDetall,
                ])->collapsible()
                ,


            ])
            ->deferLoading()
            ->paginated([50, 100, 'all']); /* I need lots of results at the same page... that adds up time. */
    }

}


```The column:

```
class CompetidorsTaulaDetall extends Column
{   
    protected string $view = 'tables.columns.competidors-taula-detall';
     private $readyToLoad = false;
    
    public function loadDetails() {
        $this->readyToLoad = true;
    }
    
    public function isReadyToLoad(){
        return $this->readyToLoad;
    }
}


```And the blade:
```
<?php
if ($isReadyToLoad()) {
/* do some ugly and slow stuff */
}
?>
<div wire:init="loadDetails" >
  @if($isReadyToLoad())
  {{-- Show something heavy. Ideally I would show a graph that takes few seconds to generate. --}}
      {{-- @livewire(\App\Filament\Widgets\AnotherHeavySlowGraphWidget::class, ['some_id' => {{ $getRecord()->some_id }} ]) --}}

  @else
  {{ __('Loading info...') }}
  @endif
</div>

```

As you see... ideally I would like to know if the blade is visible or not just before rendering, because if I am not mistaken, using alpine.js <div x-if="isCollapsed"> only toggles visibility, but does not avoid the DOM from being loaded.


DISCLAIMER:
Looking at the code I discovered I have this...  
                ->extraAttributes([
                    'wire:click' =>'$refresh',
                ])

So I can refresh the Panel at will. When I click on it, it refreshes all Panel appearances, not only the one belonging to the cell I am clicking into, but all rows. 
 
TIP: As you see, the table is defered FROM the widget, and using this trick, the Custom Columns are defered FROM the table.

**Comment by epertinez** [comment]
Ok, a follow up:

I found the way to solve the problem.

a) First of all: I think it is not possible to refresh only some livewire component instances and not the rest. When livewire sends a refresh poll event it refreshes all, not only what is being worked with.
That mean that my only way to solve the problem would be to be very, very precise on which rows are collapsed and which ones are not, to trick the system returning null to hidden rows.
But, at the same time, this wouldn't work neither. Every graph embeded in a collapsed row takes 30 seconds to create. Now, if you open 3 or 4 of those rows, every time, livewire refreshes all of them, so it takes 30seconds BY the number of opened rows. Not acceptable.

b) For what I see... alpine.js has two different attributes that seem the same but act very different...


<div x-show="open">
  ...
</div>

toggles visibility, so the DOM is already loaded. If I did use x-show="! isCollapsed" for a bunch of inserted images in my hidden rows, it wouldn't work, because internally the browser would keep asking for them, killing my server.

<template x-if="open">
  <div>...</div>
</template>

Now, x-if do not work on visibility, but what it does is to comment out the branch from the DOM tree, so it is, in fact, not in the DOM. So, in my case, if I use x-if="! isCollapsed" inside of collapsible rows, whatever is inside the template will simply not load in the dom until isCollapsed is not hidden.


With those two hints, I ended up making my customColumn load an iframe wrapped in a  x-if template:
```
<div wire:init="loadDetails" >
  @if($isReadyToLoad())
  
<template x-if="! isCollapsed">
    <div>
  <iframe src="/showMyDetail/{{$getRecord()->some_id }}" title="iframe detall"></iframe> 
    </div>
</template>
  @else
  <x-filament::loading-indicator class="h-5 w-5" /><br/>{{ __('Preparant informació per ser carregada...') }}
  @endif
</div>

```
In fact, right now it is so fast that I can get rid of all defer loading of Columns, as all columns only hold an iframe link that costs zero to generate or download.

```
<div  >
<template x-if="! isCollapsed">
    <div>
  <iframe src="/showMyDetail/{{$getRecord()->some_id }}" title="iframe detall"></iframe> 
    </div>
</template>
</div>

```

**Comment by epertinez** [comment]
I know iframe is old and ugly. Surely I have to use alpine's 


```
<div x-html="(await axios.get('/some/html/partial')).data">
  ...
</div>
```

instead.
